Miracast/MiracastPlayer/MiracastGstPlayer.cpp

```diff
@@ -0,0 +1,932 @@
+        {
+            gint percent = 0;
+            gst_message_parse_buffering(msg, &percent);
+            MIRACASTLOG_VERBOSE("Buffering [%3d%%].", percent);
+
+            /* Wait until buffering is complete before start/resume playing */
+            if (percent < 100)
+            {
+                gst_element_set_state(self->m_pipeline, GST_STATE_PAUSED);
+            }
+            else
+            {
+                gst_element_set_state(self->m_pipeline, GST_STATE_PLAYING);
+            }
+        }
+        break;
+        case GST_MESSAGE_TAG:
+        {
+            MIRACASTLOG_VERBOSE("!!!! GST_MESSAGE_TAG !!!!");
+        }
+        break;
+        case GST_MESSAGE_CLOCK_LOST:
+        {
+            MIRACASTLOG_VERBOSE("!!!! GST_MESSAGE_CLOCK_LOST !!!!");
+            /* The current clock as selected by the pipeline became unusable, then select a new clock */
+            gst_element_set_state(self->m_pipeline, GST_STATE_PAUSED);
+            gst_element_set_state(self->m_pipeline, GST_STATE_PLAYING);
+        }
+        break;
+        case GST_MESSAGE_QOS:
+        {
+            MIRACASTLOG_VERBOSE("Received [%s], a buffer was dropped or an element changed its processing strategy for Quality of Service reasons.", gst_message_type_get_name(msg->type));
+            GstFormat format;
+            guint64 processed;
+            guint64 dropped;
+            gst_message_parse_qos_stats(msg, &format, &processed, &dropped);
+            MIRACASTLOG_VERBOSE("Format [%s], Processed [%lu], Dropped [%lu].", gst_format_get_name(format), processed, dropped);
+
+            gint64 jitter;
+            gdouble proportion;
+            gint quality;
+            gst_message_parse_qos_values(msg, &jitter, &proportion, &quality);
+            MIRACASTLOG_VERBOSE("Jitter [%lu], Proportion [%lf],  Quality [%u].", jitter, proportion, quality);
+
+            gboolean live;
+            guint64 running_time;
+            guint64 stream_time;
+            guint64 timestamp;
+            guint64 duration;
+            gst_message_parse_qos(msg, &live, &running_time, &stream_time, &timestamp, &duration);
+            MIRACASTLOG_VERBOSE("live stream [%d], runninng_time [%lu], stream_time [%lu], timestamp [%lu], duration [%lu].", live, running_time, stream_time, timestamp, duration);
+        }
+        break;
+        default:
+        {
+        }
+        break;
+    }
+    MIRACASTLOG_TRACE("Exiting...\n");
+    return TRUE;
+}
+
+void MiracastGstPlayer::pad_added_handler(GstElement *gstelement, GstPad *new_pad, gpointer userdata)
+{
+    MiracastGstPlayer *self = static_cast<MiracastGstPlayer*>(userdata);
+    GstCaps *new_pad_caps = NULL;
+    GstStructure *new_pad_struct = NULL;
+
+    MIRACASTLOG_TRACE("Entering..!!!");
+
+    if (!self->m_pipeline)
+    {
+        MIRACASTLOG_ERROR("failed to link elements!");
+        gst_object_unref(self->m_pipeline);
+        self->m_pipeline = NULL;
+        MIRACASTLOG_TRACE("Exiting..!!!");
```

Comment:
combine with 671 line

Reply:
Addressed in https://github.com/rdkcentral/rdkservices/pull/4618/commits/abb5db62d2585d5e30a40d8f7c1645336732c0ea